src/pg_cron.c

```diff
@@ -906,96 +1048,259 @@ ManageCronTask(CronTask *task, TimestampTz currentTime)
+			 * do not ERROR here.  This way, we can mark the job as failed and
+			 * keep the launcher process running normally.
+			 */
+			shm_toc_initialize_estimator(&e);
+			shm_toc_estimate_chunk(&e, strlen(cronJob->database) + 1);
+			shm_toc_estimate_chunk(&e, strlen(cronJob->userName) + 1);
+			shm_toc_estimate_chunk(&e, strlen(cronJob->command) + 1);
+			shm_toc_estimate_chunk(&e, QUEUE_SIZE);
+			shm_toc_estimate_keys(&e, PG_CRON_NKEYS);
+			segsize = shm_toc_estimate(&e);
+
+			task->seg = dsm_create(segsize, DSM_CREATE_NULL_IF_MAXSEGMENTS);
+			if (task->seg == NULL)
 			{
-				/* make sure we call PQfinish on the connection */
-				task->connection = connection;
+				task->state = CRON_TASK_ERROR;
+				task->errorMessage = "unable to create a DSM segment; more "
+								"details may be available in the server log";
 
-				task->errorMessage = "connection failed";
-				task->pollingStatus = 0;
+				break;
+			}
+
+			toc = shm_toc_create(PG_CRON_MAGIC, dsm_segment_address(task->seg), segsize);
+
+			database = shm_toc_allocate(toc, strlen(cronJob->database) + 1);
+			strcpy(database, cronJob->database);
+			shm_toc_insert(toc, PG_CRON_KEY_DATABASE, database);
+
+			username = shm_toc_allocate(toc, strlen(cronJob->userName) + 1);
+			strcpy(username, cronJob->userName);
+			shm_toc_insert(toc, PG_CRON_KEY_USERNAME, username);
+
+			command = shm_toc_allocate(toc, strlen(cronJob->command) + 1);
+			strcpy(command, cronJob->command);
+			shm_toc_insert(toc, PG_CRON_KEY_COMMAND, command);
+
+			mq = shm_mq_create(shm_toc_allocate(toc, QUEUE_SIZE), QUEUE_SIZE);
+			shm_toc_insert(toc, PG_CRON_KEY_QUEUE, mq);
+			shm_mq_set_receiver(mq, MyProc);
+
+			/*
+			 * Attach the queue before launching a worker, so that we'll automatically
+			 * detach the queue if we error out.  (Otherwise, the worker might sit
+			 * there trying to write the queue long after we've gone away.)
+			 */
+			oldcontext = MemoryContextSwitchTo(TopMemoryContext);
+			MemoryContextSwitchTo(oldcontext);
+
+			/*
+			 * Prepare the background worker.
+			 *
+			 */
+			memset(&worker, 0, sizeof(BackgroundWorker));
+			worker.bgw_flags = BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+			worker.bgw_start_time = BgWorkerStart_ConsistentState;
+			worker.bgw_restart_time = BGW_NEVER_RESTART;
+			sprintf(worker.bgw_library_name, "pg_cron");
+			sprintf(worker.bgw_function_name, "CronBackgroundWorker");
+#if (PG_VERSION_NUM >= 110000)
+			snprintf(worker.bgw_type, BGW_MAXLEN, "pg_cron");
+#endif
+			snprintf(worker.bgw_name, BGW_MAXLEN, "pg_cron worker");
+			worker.bgw_main_arg = UInt32GetDatum(dsm_segment_handle(task->seg));
+			worker.bgw_notify_pid = MyProcPid;
+
+			/*
+			 * Start the worker process.
+			 */
+			if (CronLogStatement)
+			{
+				ereport(LOG, (errmsg("cron job " INT64_FORMAT " starting: %s",
+										 jobId, command)));
+			}
+			if (!RegisterDynamicBackgroundWorker(&worker, &handle))
+			{
```

Comment:
A future improvement could be to retry here if the cause of the failure is running out of max_worker_processes slots (e.g. because of a spike in parallel worker processes).